### ISSUE TITLE
[pro] Authorization issue with authorize parent_role  and gate.owner

### ISSUE THREAD

**Issue by bessey** [rmosolgo/graphql-ruby#1421]
I am finding that when `authorize: {parent_role: :blah}` is set on a field, the gate.owner for that field is `nil`:

```
#<GraphQL::Pro::Access::Gate:0x00007f9f176e84b8
 @owner=nil,
 @level=:authorize,
 @role=:read_api,
 @parent_gate=nil,
 @explicit_pundit_policy_name=nil,
 @has_pundit_policy=false,
 @pundit_policy=nil>
```

This seems surprising to me, I would expect the owner to be set, as with `authorize: :blah`, to something. In the `parent_role` case, should it not be the field's type?

**Comment by rmosolgo** [comment]
Thanks for taking the time to open these different issues (this one, as well as #1420 and #1419, and I keep #1192 in my mind) about the shortcomings of the auth framework. I'm sorry you're finding so many gotchas, how frustrating 😖 ! 

I'm not quite sure what to say -- it seems like you're implementing a pretty sophisticated auth scheme, which is great, but maybe at this point, it's appropriate to build it custom! The auth system has been a big "selling point" of graphql-pro, so I don't think it's fundamentally broken, but as you noted, it has some real limitations. 

In the long view, I'd like to share some thoughts about the future of auth in graphql-ruby, so we can at least get some gears turning. Personally, I'd like to leverage the class-based API to build a more Ruby-ish authorization system and open-source it in graphql-ruby. Something like, class methods for `visible?` and `accessible?`, and an instance method for `permitted?`, where those methods are called during execution. That would be a much more extensible and familiar system, and I hope it would help in cases like yours, where you're building something pretty advanced. I'd like to build something that takes my lessons learned from graphql-pro as well as what I've learned since starting at GitHub (where the system is very similar), and go from there.

That said, I still have a few bridges to cross before I can work more on that. In the short term, I need to get 1.8 out the door, and then probably revisit pagination and address runtime performance issues. But, I thought I would share those thoughts in case it's interesting to you!

**Comment by bessey** [comment]
Yes, I have actually ended up rolling a pretty custom auth strategy, although still on top of your auth API. And I agree, the class based API will make this all a lot easier (though I'm not sure how you would handle backwards compat with the old style? Do we just say no auth framework for proc-lovers? :P)

Since it sounds like you have an idea for a pretty new direction with auth, don't worry about me! I've got my stuff working with some workarounds, I wrote the issues up as I hit them, not really thinking about the fact this API was likely to be deprecated with the new class based system.
